### PR TITLE
frame/authority-discovery: Have authorities() return both current and next

### PR DESCRIPTION
**Pull request by mxinden**
Authority address lookups on the DHT happen periodically (every 10
mintues) and are rather slow (~10 seconds).

In order to smooth the transition period between two sessions, have the
runtime module return both the current as well as the next authority
set. Thereby the client authority module will:

1. Publish its addresses one session in advance.

2. Prefetch the addresses of authorities of the next session in advance.

---

Closes https://github.com/paritytech/substrate/issues/6772.

I am not very experienced within `/substrate/frame` thus I would very much appreciate a thorough review.

As far as I can tell this does not introduce any API or storage-format changes.

In the future we might want to return the authorities of the previous session as well, which would require storage changes in order to keep track of the previous set.

### CODE REVIEW

File: frame/authority-discovery/src/lib.rs [hunk 1]
```diff
@@ -71,17 +71,24 @@ impl<T: Trait> pallet_session::OneSessionHandler<T::AccountId> for Module<T> {
 	where
 		I: Iterator<Item = (&'a T::AccountId, Self::Key)>,
 	{
-		let keys = authorities.map(|x| x.1).collect::<Vec<_>>();
+		let mut keys = authorities.map(|x| x.1).collect::<Vec<_>>();
+		keys.sort();
+		keys.dedup();
```
Comment (kianenigma) [inline frame/authority-discovery/src/lib.rs:76]: probably harmless to keep but I don't think this is needed here.

Comment (kianenigma) [reply frame/authority-discovery/src/lib.rs:76]: Actually it is def. not needed because of 
https://github.com/paritytech/substrate/blob/549050b7f1740c90855e777daf3f9700750ad7ff/frame/session/src/lib.rs#L447-L451

File: frame/authority-discovery/src/lib.rs [hunk 2]
```diff
@@ -206,12 +214,21 @@ mod tests {
 			.map(|i| AuthorityPair::from_seed_slice(vec![i; 32].as_ref()).unwrap().public())
 			.map(AuthorityId::from)
 			.collect();
-
 		// Needed for `pallet_session::OneSessionHandler::on_new_session`.
-		let second_authorities_and_account_ids: Vec<(&AuthorityId, AuthorityId)> = second_authorities.clone()
+		let second_authorities_and_account_ids = second_authorities.clone()
```
Comment (kianenigma) [inline frame/authority-discovery/src/lib.rs:218]: Should technically have the same set as the first one, but anyhow fine 

https://github.com/paritytech/substrate/blob/549050b7f1740c90855e777daf3f9700750ad7ff/frame/session/src/lib.rs#L225-L229

File: frame/authority-discovery/src/lib.rs [hunk 3]
```diff
@@ -233,23 +250,51 @@ mod tests {
 			AuthorityDiscovery::on_genesis_session(
 				first_authorities.iter().map(|id| (id, id.clone()))
 			);
+			first_authorities.sort();
```
Comment (kianenigma) [inline frame/authority-discovery/src/lib.rs:253]: ?

File: frame/authority-discovery/src/lib.rs (hunk 1 above)
Comment (mxinden) [reply frame/authority-discovery/src/lib.rs:76]: Thanks for the link! Will remove these lines.

File: frame/authority-discovery/src/lib.rs [hunk 4]
```diff
@@ -71,17 +71,20 @@ impl<T: Trait> pallet_session::OneSessionHandler<T::AccountId> for Module<T> {
...
 	{
-		// Remember who the authorities are for the new session.
+		// Remember who the authorities are for the new and next session.
 		if changed {
-			Keys::put(validators.map(|x| x.1).collect::<Vec<_>>());
+			let mut keys = validators.chain(queued_validators).map(|x| x.1).collect::<Vec<_>>();
+			keys.sort();
+			keys.dedup();
+
+			Keys::put(keys);
```
Comment (bkchr) [inline frame/authority-discovery/src/lib.rs:87]:
```suggestion
			let keys = validators.chain(queued_validators).map(|x| x.1).collect::<BTreeSet<_>>();

			Keys::put(keys.into_iter().collect::<Vec<_>>());
```

Probably faster?

**Comment by mxinden** [comment]
Putting this on-ice for now, giving more time to investigate https://github.com/paritytech/polkadot/issues/1544.

**Comment by gavofyork** [comment]
> Authority address lookups on the DHT happen periodically (every 10 mintues) and are rather slow (~10 seconds).

Why not increase the regularity (e.g. to every minute) and then reduce the weight (~1 second)?

**Comment by mxinden** [comment]
> > Authority address lookups on the DHT happen periodically (every 10 mintues) and are rather slow (~10 seconds).
> 
> Why not increase the regularity (e.g. to every minute) and then reduce the weight (~1 second)?

@gavofyork I don't understand what you mean with _weight_ above. Could you rephrase the sentence? Details on how long a DHT query takes today can be found in https://github.com/paritytech/polkadot/issues/1544#issuecomment-674077888.

Looking up 500 authorities on the DHT roughly requires opening 2_000 TCP connections. While not problematic in itself, I don't see the benefit of doing so every minute instead of every 10 minutes.

**Comment by tomaka** [comment]
> Putting this on-ice for now, giving more time to investigate paritytech/polkadot#1544.

Do we actually need to investigate paritytech/polkadot#1544 first?
Sooner or later we will need to merge this PR, and its fate cannot depend on whether we can keep the number of TCP connections low.
I also don't think it makes paritytech/polkadot#1544 more difficult to investigate, does it?

**Comment by gavofyork** [comment]
> > > Authority address lookups on the DHT happen periodically (every 10 mintues) and are rather slow (~10 seconds).
> > 
> > 
> > Why not increase the regularity (e.g. to every minute) and then reduce the weight (~1 second)?
> 
> @gavofyork I don't understand what you mean with _weight_ above. Could you rephrase the sentence? Details on how long a DHT query takes today can be found in [paritytech/polkadot#1544 (comment)](https://github.com/paritytech/polkadot/issues/1544#issuecomment-674077888).
> 
> Looking up 500 authorities on the DHT roughly requires opening 2_000 TCP connections. While not problematic in itself, I don't see the benefit of doing so every minute instead of every 10 minutes.

Weight == resources used. Comparatively lots of resources -> heavy operation, comparatively few -> light operation.

We should aim to minimise spikes in resource usage. Therefore 1000 connections every 5 minutes is prefereable to 2000 connections every 10 minutes. 10 connections every 3 seconds is better still.

**Comment by mxinden** [comment]
> We should aim to minimise spikes in resource usage, therefore frequently-but-lightly using resources is better than sporadic-but-heavy use of resources. The reason is that if it ever happens that two heavy-and-sporadic processes get scheduled at the same time, it can lead to resource starvation of each other or other processes which could lead to a potential attack vector. The more things are spread out over time, the more stable the system.

That makes sense. Thanks @gavofyork for the detailed explanation. As far as I can tell, the current operation of authority discovery does not introduce CPU or memory spikes (see graph below). We might expand the set of authorities to lookup in the future. I have added https://github.com/paritytech/substrate/issues/6910#issuecomment-679143425 to make sure we revisit the assumption of low resource usage.

image

**Comment by mxinden** [comment]
Removing the `onice` label.

> paritytech-ci added the D2-breaksapi label 24 days ago

As far as I can tell this pull request does not break any APIs. Can someone confirm that?

**Comment by mxinden** [comment]
> > paritytech-ci added the D2-breaksapi label 24 days ago
> 
> As far as I can tell this pull request does not break any APIs. Can someone confirm that?

@kianenigma @bkchr can one of you confirm this?

**Comment by kianenigma** [comment]
Last time I knew the label was used for PRs that need to bump the spec version. I can't see that happening here. 99.99% sure it is false positive and can ignore, but can you please confirm when this label is added @s3krit?

**Comment by s3krit** [comment]
The relevant logic for check_runtime seems to be as follows:

If there have been changes to any files in `bin/node/src/runtime/*`, `frame/*` or `primitives/sr-*` between this commit and `master`, AND the `spec_version` (specified in `bin/node/runtime/src/lib.rs`) has changed since the last release (checked using the rolling `release` tag), apply the 'D2-breaksapi' label and exit successfully.

I'm not quite sure of the reasoning behind applying that label to be honest. Perhaps that is erroneously there and should be shifted to the next clause in the script (which runs if the spec_version and impl_version *hasn't* been bumped - and fails the job because of the need to bump either spec_version or impl_version if there have been relevant changes).

**Comment by mxinden** [comment]
> The relevant logic for check_runtime seems to be as follows:
> 
> If there have been changes to any files in `bin/node/src/runtime/*`, `frame/*` or `primitives/sr-*` between this commit and `master`, AND the `spec_version` (specified in `bin/node/runtime/src/lib.rs`) has changed since the last release (checked using the rolling `release` tag), apply the 'D2-breaksapi' label and exit successfully.
> 
> I'm not quite sure of the reasoning behind applying that label in that situation to be honest. Perhaps that is erroneously there and should be shifted to the next clause in the script (which runs if the spec_version and impl_version _hasn't_ been bumped - and fails the job because of the need to bump either spec_version or impl_version if there have been relevant changes).

@s3krit @kianenigma do I understand correctly that the `D2-breaksapi` label can be ignored and this pull request can be merged as is without any migration or adapter logic?

**Comment by s3krit** [comment]
That is my understanding yes (just based on following the logic of check_runtime) but you might want to wait for confirmation from @kianenigma

**Comment by kianenigma** [comment]
Yes it is fine. There are no storage changes and the change to the runtime API is not to the signature, so all good.

**Comment by mxinden** [comment]
bot merge

**Comment by mxinden** [comment]
Thanks for the help everyone!

**Comment by gavofyork** [comment]
> resource

By "resource" I was not merely refereing to CPU or memory. TCP connections will tend to be in limited supply on a system and come with their own overheads and still count as a "resource".

**Comment by mxinden** [comment]
> The graph makes it clear that there are spikes in this resource, which we should work to reduce.

This is addressed in https://github.com/paritytech/substrate/pull/7018. See https://github.com/paritytech/substrate/pull/7018#issuecomment-687043009 for an analysis on its resource consumption. @gavofyork let me know what you think.
